Improve PR template checklist (#9429)
gfx-rs/wgpu: .github/pull_request_template.md

**Checklist**

<!-- Note that checking all the boxes is not necessary to open a PR. -->

- [ ] I self-reviewed this PR and fully understand it.

> - [ ] This PR potentially affects behavior on WebGPU.

Reviewer: By "behavior on WebGPU", do you mean "behavior of `Backend::BrowserWebGpu`"? If so, I’m not sure why this specifically is worth a checklist point where other cases are not. Or do you mean “behavior <ins>of `wgpu` as a</ins> WebGPU implementation”?

Author: This is worth noting for Mozilla review purposes. If a PR is entirely a bug fix within a native feature that already has proper validation in place, firefox does not have to care about it, is my understanding.

Reviewer: Okay, so you mean the latter (behavior of `wgpu` as a WebGPU implementation).
```suggestion
- [ ] This PR potentially affects the behavior of `wgpu` used as a WebGPU implementation.
```

Author: how about
```suggestion
- [ ] This PR potentially affects WebGPU implementation behavior.
```

Reviewer: No, that does not clarify the ambiguity at all. `wgpu` plays *two* roles in relationship to WebGPU:

1. It provides a Rust API implemented on top of an existing WebGPU API.
2. It provides some ingredients with which one can build a WebGPU API implementation.

For this to be clear, **it must specify which of the two is meant.**

Author: How about
```suggestion
- [ ] WebGPU implementations built with `wgpu` may be affected behaviorally.
```

Reviewer: That seems clear enough to me. I don’t like “affected behaviorally” because I think the world needs fewer adjective-or-adverb-pileups and more actual sentence structure, so I’d choose

```suggestion
- [ ] The behavior of WebGPU implementations built with `wgpu` may be affected.
```

but I don’t insist.

> - [ ] I added all necessary validation to ensure behavior changes are confined to where they should be.

Reviewer: I’m not sure how "validation" and "behavior changes" are related here. Are you using "validation" to mean something related to testing/QA? If so, I'm not confidently familiar with this as a precise term, and I don’t think other readers will be either.

In the context of the wgpu project, "validation" usually means "checks on the input performed by wgpu before things reach the underlying API" (or similar checks performed by Vulkan, etc). If this is what you mean, then I am confused about what it means to “confine behavior changes” via validation.

Author: I'm saying that we shouldn't be allowing new behavior to be observable (like mesh shader logic) if the feature is not enabled (it should just error on any attempts to use it)

Reviewer: That makes a lot of sense but it was not at all clear. How about something like:

> * [ ] All public API changes (including behavior, not just signatures) are intentional, documented, and have appropriate validation.
>     * [ ] New functionality should be gated by `Features` as appropriate.

Author: I don't like having several items (intentional, documented, and have appropriate validation) as one checkbox. The intentional should be covered by the `I self-reviewed this PR and fully understand it`, and the documentation merits its own checkbox. Validation and documentation are totally orthogonal things imo and shouldnt be conflated

The second checkbox you proposed doesnt read like a claim, it reads like a command, maybe it should be in `<!-- -->` above?

Author: The docs one could be combined with the last checkbox about relevant context.

Reviewer: I'm not particularly attached to the specific wording I proposed. I just think that this needs to be more explicit about (1) what you mean by "validation", and (2) how this relates to "behavior changes".

Reviewer 2: I agree that this point needs to be more clear, how to get there, I don't currently have an opinion unless you need me to have one. I will caution against too many checkboxes, as it will decrease the chance of people actually reading it, so combining mildly related things in one box is a good idea.

Author: I fear longer checkbox messages more than too many checkboxes

  - [ ] I added validation tests demonstrating that any behavior changes are in fact confined with proper error messages.

> - [ ] I added all necessary `CHANGELOG.md` entries for this change. <!-- See instructions at the top of `CHANGELOG.md`. -->

Reviewer: ```suggestion
- [ ] I added all necessary `CHANGELOG.md` entries for the user-facing effects of this change. <!-- See instructions at the top of `CHANGELOG.md`. -->
```

Author: I was hedging this with `all necessary`, how about
```suggestion
- [ ] I added `CHANGELOG.md` entries for the user-facing effects of this change. <!-- See instructions at the top of `CHANGELOG.md`. -->
```

Reviewer: Sounds good.

- [ ] I think this PR is a minimal change that doesn't make sense to land as multiple separate PRs.
- [ ] I think the commit history is logical and decently reviewable.

> - [ ] The PR description contains enough context for a reviewer to understand the motivation for this PR, and the solution it implements.

Author: Here's a second pass on what it could look like
```suggestion
- [ ] I self-reviewed and fully understand this PR.
- [ ] WebGPU implementations built with `wgpu` may be affected behaviorally.
- [ ] Validation and feature gates are in place to confine behavioral changes.
- [ ] Tests demonstrate the validation and altered logic works. <!-- See `docs/testing.md` -->
- [ ] `CHANGELOG.md` entries for the user-facing effects of this change are present. <!-- See instructions at the top of `CHANGELOG.md`. -->
- [ ] The PR is minimal, and doesn't make sense to land as multiple PRs.
- [ ] Commits are logically scoped and individually reviewable.
- [ ] The PR description has enough context to understand the motivation and solution implemented.
```

The idea is to have the most important words of the sentence as close to the start as possible, so you can scan the list more easily. I also golfed it a bit to make it less wordy

Rendered:

- [ ] I self-reviewed and fully understand this PR.
- [ ] WebGPU implementations built with `wgpu` may be affected behaviorally.
- [ ] Validation and feature gates are in place to confine behavioral changes.
- [ ] Tests demonstrate the validation and altered logic works. <!-- See `docs/testing.md` -->
- [ ] `CHANGELOG.md` entries for the user-facing effects of this change are present. <!-- See instructions at the top of `CHANGELOG.md`. -->
- [ ] The PR is minimal, and doesn't make sense to land as multiple PRs.
- [ ] Commits are logically scoped and individually reviewable.
- [ ] The PR description has enough context to understand the motivation and solution implemented.

Reviewer: Looks good to me.
